[PyTorch][2/4] Support static dispatch with multiple backends

priyaramani opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75605

Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

priyaramani committed:
[PyTorch][2/4] Support static dispatch with multiple backends

Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

[ghstack-poisoned]

priyaramani committed:
Update on "[PyTorch][2/4] Support static dispatch with multiple backends"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35510644).

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -205,12 +206,11 @@ def cpp_string(s: str) -> str:
-def static_dispatch_keys(backend: Optional[BackendIndex]) -> List[DispatchKey]:
-    if backend is None:
+def static_dispatch_keys(backends: List[BackendIndex]) -> List[DispatchKey]:
+    if backends is None:
```
nit: you can probably remove the none check now that the input isn't `Optional` (passing in `None` should fail mypy)

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -231,41 +231,80 @@ def get_static_dispatch_backend(f: NativeFunction, backend_index: BackendIndex)
+        *, method: bool, backend_indices: List[BackendIndex]
 ) -> Optional[str]:
-    if backend_index is None or f.manual_kernel_registration:
+    if backend_indices is None or f.manual_kernel_registration:
```
you should also be able to remove this `None` check

bdhirsh replied:
oh, it should probably be a check for `len(backend_indices) == 0`? And the codegen should default it to the empty list, instead of none?

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -231,41 +231,80 @@ def get_static_dispatch_backend(f: NativeFunction, backend_index: BackendIndex)
+    elif len(keys) == 0:
+        return generate_static_dispatch(f, cpp_sig, method=method, backend_index=backend_indices[0])
+    else:
+        return f'TORCH_CHECK(false, "Static dispatch does not support backends with multiple kernels");'
```
Hmm... if you pass in two backends to the codegen (say `CPU` and `QuantizedCPU`), then won't the codegen emit this torch check for every `f.structured_delegate` NativeFunction?

priyaramani replied:
The else part is implemented later in the stack https://github.com/pytorch/pytorch/pull/75607. Split them up to keep diff sizes manageable. 
Also if we pass in multiple backends for a native function, and if they have a structured_delegate and one or more kernels for the same, then this torch check will be emitted as of this diff.

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -231,41 +231,80 @@ def get_static_dispatch_backend(f: NativeFunction, backend_index: BackendIndex)
+    if len(keys) == 1:
+        return generate_static_dispatch(f, cpp_sig, method=method, backend_index=keys[0])
+    elif len(keys) == 0:
+        return generate_static_dispatch(f, cpp_sig, method=method, backend_index=backend_indices[0])
```
What's the idea behind passing `backend_indices[0]` here?

priyaramani replied:
@larryliu0820 
If none of the static dispatch backends have kernels for the native function, then we just pass in the first one in the list, could be any one of them, as in generate_static_dispatch(f, ..) if the backend doesn't have a kernel or if the native function doesn't have a structured delegate we only see if the native function has compositeImplicitAugradKernel or compositeImplicitAugradKernel and use that if not emit a TORCH_CHECK https://github.com/pytorch/pytorch/pull/75605/files#diff-6382c847dac2d011f822c492e61f9fa0eceb47c2d6fc5fb4f0c42af726c1a3aaR283-R288

priyaramani committed:
Update on "[PyTorch][2/4] Support static dispatch with multiple backends"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

[ghstack-poisoned]

priyaramani committed:
Update on "[PyTorch][2/4] Support static dispatch with multiple backends"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

[ghstack-poisoned]

priyaramani committed:
Update on "[PyTorch][2/4] Support static dispatch with multiple backends"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

[ghstack-poisoned]

priyaramani committed:
Update on "[PyTorch][2/4] Support static dispatch with multiple backends"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

[ghstack-poisoned]

larryliu0820 reviewed (approved):
Is it possible to add a unit test before commit? Just to demonstrate the ability we are adding to static dispatch. You can find two operators with different kernels and write a small model.

priyaramani committed:
Update on "[PyTorch][2/4] Support static dispatch with multiple backends"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

[ghstack-poisoned]

priyaramani committed:
Update on "[PyTorch][2/4] Support static dispatch with multiple backends"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

[ghstack-poisoned]

priyaramani committed:
Update on "[PyTorch][2/4] Support static dispatch with multiple backends"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

This change passes in List[BackendIndex] and adds ability to generate code for multiple static backends with 1 or 0 kernels

Differential Revision: [D35510644](https://our.internmc.facebook.com/intern/diff/D35510644/)

[ghstack-poisoned]
